data/tutorials/language/0lg_00_basic_datatypes.md

```diff
@@ -0,0 +1,844 @@
+  | Rogue
+  | Sorcerer
+  | Wizard
+
+# type rectitude = Evil | R_Neutral | Good;;
+type rectitude = Evil | R_Neutral | Good
+
+# type firmness = Chaotic | F_Neutral | Lawful;;
+type firmness = Chaotic | F_Neutral | Lawful
+```
+
+These kinds of variant types can also be used to represent weekdays, cardinal
+directions, or any other fixed-sized set of values that can be given names. An
+ordering is defined on values following the definition order (e.g., `Druid
+< Ranger`).
+
+Here is how pattern matching can be performed on the types defined above.
+```ocaml
+# let rectitude_to_french = function
+    | Evil -> "Mauvais"
+    | R_Neutral -> "Neutre"
+    | Good -> "Bon";;
+val rectitude_to_french : rectitude -> string = <fun>
+```
+
+Note that:
+
+- `unit` is a variant with a unique constructor, which does not carry data: `()`.
+- `bool` is also a variant with two constructors that don't carry data: `true` and `false`.
+
+A pair `(x, y)` has type `a * b`, where `a` is the type of `x` and `b` is the type of `y`. Some may find it intriguing that `a * b` is called a product. Although this is not a complete explanation, here is a remark that may help in understanding: Consider the product type `character_class * character_alignement`. There are 12 classes and 9 alignments. Any pair of values from those types inhabits the product type. Therefore, in the product type, there are 9 × 12 = 108 values, which is also a product.
+
+#### Constructors With Data
+
+It is possible to wrap data in constructors. The following type has several constructors with data (e.g., `Hash of string`) and some without (e.g., `Head`). It represents the different means to refer to a Git [revision](https://git-scm.com/docs/gitrevisions).
+```ocaml
+# type commit =
+  | Hash of string
+  | Tag of string
+  | Branch of string
+  | Head
+  | Fetch_head
+  | Orig_head
+  | Merge_head;;
+type commit =
+    Hash of string
+  | Tag of string
+  | Branch of string
+  | Head
+  | Fetch_head
+  | Orig_head
+  | Merge_head
+
+```
+
+Here is how to convert a `commit` to a `string` using pattern matching:
+```ocaml
+# let commit_to_string = function
+  | Hash sha -> sha
+  | Tag name -> name
+  | Branch name -> name
+  | Head -> "HEAD"
+  | Fetch_head -> "FETCH_HEAD"
+  | Orig_head -> "ORIG_HEAD"
+  | Merge_head -> "MERGE_HEAD";;
+val commit_to_string : commit -> string = <fun>
+```
+
+Here, the `function …` construct is used instead of the `match … with …` construct used previously.
+```ocaml
+let commit_to_string' x = match x with
+  | Hash sha -> sha
+  | Tag name -> name
+  | Branch name -> name
+  | Head -> "HEAD"
+  | Fetch_head -> "FETCH_HEAD"
+  | Orig_head -> "ORIG_HEAD"
+  | Merge_head -> "MERGE_HEAD";;
+val commit_to_string' : commit -> string = <fun>
+```
+
+The `match … with …` construct needs to be passed an expression that is inspected. The `function …` is a special form of anonymous function taking a parameter and forwarding it to a `match … with …` construct as shown above.
```

Comment:
```suggestion
We need to pass an inspected expression to the `match … with …` construct. The `function …` is a special form of anonymous function that takes a parameter and forwards it to a `match … with …` construct, as shown above.
```

Does this first sentence still say the same thing?

Reply:
Indeed, except in much better way.